From ba4307dfaf8889e9726c08e718ee43e212501447 Mon Sep 17 00:00:00 2001
From: Konda Reddy Kachana <kkachana@google.com>
Date: Tue, 14 Mar 2023 06:03:06 +0000
Subject: [PATCH 1/4] bmcweb: Add RDE Device aggregation support

Extend the Redfish aggregation to support RDE devices.

Tested:
Queries to a collection resulted all the satellite resources
correctly. Also, query to a aggregated resource worked
correctly.

Upstream-Status: Inappropriate [oe-specific]
Change-Id: Ie7f7a99dda283d5d3ac61964aa9e5833cf3a02aa
Signed-off-by: Konda Reddy Kachana <kkachana@google.com>
---
 redfish-core/include/redfish_aggregator.hpp | 302 +++++++++++++++++++-
 redfish-core/lib/aggregation_service.hpp    |   8 +-
 2 files changed, 293 insertions(+), 17 deletions(-)

diff --git a/redfish-core/include/redfish_aggregator.hpp b/redfish-core/include/redfish_aggregator.hpp
index 0a4ea4a7..5112747b 100644
--- a/redfish-core/include/redfish_aggregator.hpp
+++ b/redfish-core/include/redfish_aggregator.hpp
@@ -29,6 +29,10 @@ enum class SearchType
     ContainsSubordinate,
     Resource
 };
+struct RdeSatelliteConfig
+{
+    std::string name;
+};
 
 // clang-format off
 // These are all of the properties as of version 2022.2 of the Redfish Resource
@@ -388,7 +392,9 @@ class RedfishAggregator
     // of satellite configs when the class is first created
     static void constructorCallback(
         const boost::system::error_code& ec,
-        const std::unordered_map<std::string, boost::urls::url>& satelliteInfo)
+        const std::unordered_map<std::string, boost::urls::url>& satelliteInfo,
+        const std::unordered_map<std::string, RdeSatelliteConfig>&
+            rdeSatelliteInfo)
     {
         if (ec)
         {
@@ -398,13 +404,16 @@ class RedfishAggregator
 
         BMCWEB_LOG_DEBUG("There were {} satellite configs found at startup",
                          std::to_string(satelliteInfo.size()));
+        BMCWEB_LOG_DEBUG("There were {} RDE Device configs found at startup",
+                         std::to_string(rdeSatelliteInfo.size()));
     }
 
     // Search D-Bus objects for satellite config objects and add their
     // information if valid
     static void findSatelliteConfigs(
         const dbus::utility::ManagedObjectType& objects,
-        std::unordered_map<std::string, boost::urls::url>& satelliteInfo)
+        std::unordered_map<std::string, boost::urls::url>& satelliteInfo,
+        std::unordered_map<std::string, RdeSatelliteConfig>& rdeSatelliteInfo)
     {
         for (const auto& objectPath : objects)
         {
@@ -430,6 +439,14 @@ class RedfishAggregator
                     addSatelliteConfig("5B247A", interface.second,
                                        satelliteInfo);
                 }
+                else if (
+                    interface.first ==
+                    "xyz.openbmc_project.Configuration.RdeSatelliteController")
+                {
+                    BMCWEB_LOG_DEBUG("Found RDE Satellite Controller at {}",
+                                     objectPath.first.str);
+                    addRdeSatelliteConfig(interface.second, rdeSatelliteInfo);
+                }
             }
         }
     }
@@ -532,6 +549,44 @@ class RedfishAggregator
                          result.first->second.encoded_host_and_port());
     }
 
+    // Parse the properties of a RDE Device config object and add the
+    // configuration if the properties are valid
+    static void addRdeSatelliteConfig(
+        const dbus::utility::DBusPropertiesMap& properties,
+        std::unordered_map<std::string, RdeSatelliteConfig>& rdeSatelliteInfo)
+    {
+        RdeSatelliteConfig rdeConfig;
+        std::string name;
+        for (const auto& prop : properties)
+        {
+            if (prop.first == "Name")
+            {
+                const std::string* propVal =
+                    std::get_if<std::string>(&prop.second);
+                if (propVal == nullptr)
+                {
+                    BMCWEB_LOG_ERROR("Invalid Name value");
+                    return;
+                }
+                rdeConfig.name = *propVal;
+            }
+        } // Finished reading properties
+
+        // Set the prefix to a random string 'E0SB8D'
+        // (TODO) Generate a unique random prefix for each RDE Device
+        name = "E0SB8D";
+        std::string resultString;
+        auto result = rdeSatelliteInfo.emplace(name, std::move(rdeConfig));
+        if (result.second)
+        {
+            resultString = "Added new RDE Device config ";
+        }
+        else
+        {
+            resultString = "Updated existing RDE Device config ";
+        }
+    }
+
     enum AggregationType
     {
         Collection,
@@ -581,8 +636,11 @@ class RedfishAggregator
         const crow::Request& req,
         const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
         const std::unordered_map<std::string, boost::urls::url>& satelliteInfo,
+        const std::unordered_map<std::string, RdeSatelliteConfig>&
+            rdeSatelliteInfo,
         std::string_view memberName)
     {
+        bool validPrefix = false;
         // Determine if the resource ID begins with a known prefix
         for (const auto& satellite : satelliteInfo)
         {
@@ -596,10 +654,29 @@ class RedfishAggregator
                 // then forward to the associated satellite BMC
                 getInstance().forwardRequest(req, asyncResp, satellite.first,
                                              satelliteInfo);
-                return;
+                validPrefix = true;
             }
         }
-
+        // Determine if the resource ID begins with a known prefix
+        for (const auto& rdeSatellite : rdeSatelliteInfo)
+        {
+            std::string targetPrefix = rdeSatellite.first;
+            targetPrefix += "_";
+            if (memberName.starts_with(targetPrefix))
+            {
+                BMCWEB_LOG_DEBUG("\"{}\" is a known prefix",
+                                 rdeSatellite.first);
+                // Remove the known prefix from the request's URI and
+                // then forward to RDE Daemon
+                getInstance().forwardRdeRequest(req, asyncResp, rdeSatellite.first,
+                                  rdeSatelliteInfo);
+                validPrefix = true;
+            }
+        }
+        if (validPrefix)
+        {
+            return;
+        }
         // We didn't recognize the prefix and need to return a 404
         std::string nameStr = req.url().segments().back();
         messages::resourceNotFound(asyncResp->res, "", nameStr);
@@ -612,7 +689,9 @@ class RedfishAggregator
         const std::shared_ptr<crow::Request>& sharedReq,
         const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
         const boost::system::error_code& ec,
-        const std::unordered_map<std::string, boost::urls::url>& satelliteInfo)
+        const std::unordered_map<std::string, boost::urls::url>& satelliteInfo,
+        const std::unordered_map<std::string, RdeSatelliteConfig>&
+            rdeSatelliteInfo)
     {
         if (sharedReq == nullptr)
         {
@@ -625,9 +704,9 @@ class RedfishAggregator
             return;
         }
 
-        // No satellite configs means we don't need to keep attempting to
+        // No satellite or RDE configs means we don't need to keep attempting to
         // aggregate
-        if (satelliteInfo.empty())
+        if (satelliteInfo.empty() && rdeSatelliteInfo.empty())
         {
             // For collections or resources that can contain a subordinate
             // top level collection we'll also handle the request locally so we
@@ -651,8 +730,8 @@ class RedfishAggregator
             BMCWEB_LOG_DEBUG("Aggregating a collection");
             // We need to use a specific response handler and send the
             // request to all known satellites
-            getInstance().forwardCollectionRequests(thisReq, asyncResp,
-                                                    satelliteInfo);
+            getInstance().forwardCollectionRequests(
+                thisReq, asyncResp, satelliteInfo, rdeSatelliteInfo);
             return;
         }
 
@@ -684,7 +763,8 @@ class RedfishAggregator
             {
                 // We've matched a resource collection so this current segment
                 // must contain an aggregation prefix
-                findSatellite(thisReq, asyncResp, satelliteInfo, *it);
+                findSatellite(thisReq, asyncResp, satelliteInfo,
+                              rdeSatelliteInfo, *it);
                 return;
             }
 
@@ -746,7 +826,9 @@ class RedfishAggregator
     void forwardCollectionRequests(
         const crow::Request& thisReq,
         const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
-        const std::unordered_map<std::string, boost::urls::url>& satelliteInfo)
+        const std::unordered_map<std::string, boost::urls::url>& satelliteInfo,
+        const std::unordered_map<std::string, RdeSatelliteConfig>&
+            rdeSatelliteInfo)
     {
         for (const auto& sat : satelliteInfo)
         {
@@ -764,8 +846,189 @@ class RedfishAggregator
                                         ensuressl::VerifyCertificate::Verify,
                                         thisReq.fields(), thisReq.method(), cb);
         }
+        for (const auto& rsat : rdeSatelliteInfo)
+        {
+            BMCWEB_LOG_DEBUG(" Collection Request: dbus call to RDE Daemon");
+            std::string targetURI(thisReq.target());
+
+            crow::connections::systemBus->async_method_call(
+                [rsat, asyncResp](const boost::system::error_code ec,
+                                  const std::string& jsonString) {
+                if (ec)
+                {
+                    BMCWEB_LOG_ERROR("DBUS response error {}, {}", ec.value(),
+                                     ec.message());
+                    return;
+                }
+                processRdeCollectionResponse(rsat.first, asyncResp, jsonString);
+                },
+                "xyz.openbmc_project.rdeService",
+                "/xyz/openbmc_project/rdeObject",
+                "xyz.openbmc_project.rdeInterface", "getCollection", targetURI);
+        }
     }
 
+    // Attempt to forward a request to the RDE Daemon associated with the
+    // prefix.
+    static void forwardRdeRequest(
+        const crow::Request& thisReq,
+        const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+        const std::string& prefix,
+        const std::unordered_map<std::string, RdeSatelliteConfig>&
+            rdeSatelliteInfo)
+    {
+        const auto& sat = rdeSatelliteInfo.find(prefix);
+        if (sat == rdeSatelliteInfo.end())
+        {
+            // Realistically this shouldn't get called since we perform an
+            // earlier check to make sure the prefix exists
+            BMCWEB_LOG_ERROR("Unrecognized RDE Device prefix \"{}\"", prefix);
+            return;
+        }
+
+        // We need to strip the prefix from the request's path
+        std::string targetURI(thisReq.target());
+        size_t pos = targetURI.find(prefix + "_");
+        if (pos == std::string::npos)
+        {
+            // If this fails then something went wrong
+            BMCWEB_LOG_ERROR("Error removing prefix \"{}_\" from request URI",
+                             prefix);
+            messages::internalError(asyncResp->res);
+            return;
+        }
+        targetURI.erase(pos, prefix.size() + 1);
+
+        BMCWEB_LOG_DEBUG(" Resource Request: dbus call to RDE Daemon");
+
+        crow::connections::systemBus->async_method_call(
+            [prefix, asyncResp](const boost::system::error_code ec,
+                                const std::string& jsonString) {
+            if (ec)
+            {
+                BMCWEB_LOG_ERROR("DBUS response error {}, {}", ec.value(),
+                                 ec.message());
+                return;
+            }
+            processRdeResponse(prefix, asyncResp, jsonString);
+            },
+            "xyz.openbmc_project.rdeService", "/xyz/openbmc_project/rdeObject",
+            "xyz.openbmc_project.rdeInterface", "getResource", targetURI);
+    }
+    // Processes the response returned by a RDE Device and loads its
+    // contents into asyncResp
+    static void
+        processRdeResponse(std::string_view prefix,
+                           const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                           const std::string& respString)
+    {
+        nlohmann::json jsonVal = nlohmann::json::parse(respString, nullptr,
+                                                       false);
+        if (jsonVal.is_discarded())
+        {
+            BMCWEB_LOG_ERROR("Error parsing RDE Device response as JSON");
+            messages::operationFailed(asyncResp->res);
+            return;
+        }
+
+        BMCWEB_LOG_DEBUG("Successfully parsed RDE Device response");
+
+        // TODO: For collections we  want to add the satellite responses to
+        // our response rather than just straight overwriting them if our
+        // local handling was successful (i.e. would return a 200).
+        addPrefixes(jsonVal, prefix);
+
+        BMCWEB_LOG_DEBUG("Added prefix to parsed RDE Device response");
+
+        asyncResp->res.result(200);
+        asyncResp->res.jsonValue = std::move(jsonVal);
+
+        BMCWEB_LOG_DEBUG("Finished writing asyncResp");
+    }
+
+    // Processes the collection response returned by a RDE Device and merges
+    // its "@odata.id" values
+    static void processRdeCollectionResponse(
+        const std::string& prefix,
+        const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+        const std::string& respString)
+    {
+        nlohmann::json jsonVal = nlohmann::json::parse(respString, nullptr,
+                                                       false);
+        if (jsonVal.is_discarded())
+        {
+            BMCWEB_LOG_ERROR("Error parsing RDEd response as JSON");
+
+            // Notify the user if doing so won't overwrite a valid response
+            if ((asyncResp->res.resultInt() != 200) &&
+                (asyncResp->res.resultInt() != 502))
+            {
+                messages::operationFailed(asyncResp->res);
+            }
+            return;
+        }
+
+        BMCWEB_LOG_DEBUG("Successfully parsed RDEd response");
+
+        // Now we need to add the prefix to the URIs contained in the
+        // response.
+        addPrefixes(jsonVal, prefix);
+
+        BMCWEB_LOG_DEBUG("Added prefix to parsed RDE Device response");
+
+        // If this resource collection does not exist on the aggregating bmc
+        // and has not already been added from processing the response from
+        // a different satellite then we need to completely overwrite
+        // asyncResp
+        if (asyncResp->res.resultInt() != 200)
+        {
+            // We only want to aggregate collections that contain a
+            // "Members" array
+            if ((!jsonVal.contains("Members")) &&
+                (!jsonVal["Members"].is_array()))
+            {
+                BMCWEB_LOG_DEBUG("Skipping aggregating unsupported resource");
+                return;
+            }
+
+            BMCWEB_LOG_DEBUG(
+                "Collection does not exist, overwriting asyncResp");
+            asyncResp->res.jsonValue = std::move(jsonVal);
+
+            BMCWEB_LOG_DEBUG("Finished overwriting asyncResp");
+        }
+        else
+        {
+            // We only want to aggregate collections that contain a
+            // "Members" array
+            if ((!asyncResp->res.jsonValue.contains("Members")) &&
+                (!asyncResp->res.jsonValue["Members"].is_array()))
+
+            {
+                BMCWEB_LOG_DEBUG("Skipping aggregating unsupported resource");
+                return;
+            }
+
+            BMCWEB_LOG_DEBUG(
+                "Adding aggregated resources from \"{}\" to collection",
+                prefix);
+
+            // TODO: This is a potential race condition with multiple
+            // satellites and the aggregating bmc attempting to write to
+            // update this array.  May need to cascade calls to the next
+            // satellite at the end of this function.
+            auto& members = asyncResp->res.jsonValue["Members"];
+            auto& satMembers = jsonVal["Members"];
+            for (auto& satMem : satMembers)
+            {
+                members.push_back(std::move(satMem));
+            }
+            asyncResp->res.jsonValue["Members@odata.count"] = members.size();
+
+            // TODO: Do we need to sort() after updating the array?
+        }
+    } // End processRdeCollectionResponse()
+
     // Forward request for a URI that is uptree of a top level collection to
     // each known satellite BMC
     void forwardContainsSubordinateRequests(
@@ -817,7 +1080,8 @@ class RedfishAggregator
     static void getSatelliteConfigs(
         std::function<
             void(const boost::system::error_code&,
-                 const std::unordered_map<std::string, boost::urls::url>&)>
+                 const std::unordered_map<std::string, boost::urls::url>&,
+                 const std::unordered_map<std::string, RdeSatelliteConfig>&)>
             handler)
     {
         BMCWEB_LOG_DEBUG("Gathering satellite configs");
@@ -828,18 +1092,20 @@ class RedfishAggregator
                 const boost::system::error_code& ec,
                 const dbus::utility::ManagedObjectType& objects) {
                 std::unordered_map<std::string, boost::urls::url> satelliteInfo;
+                std::unordered_map<std::string, RdeSatelliteConfig>
+                    rdeSatelliteInfo;
                 if (ec)
                 {
                     BMCWEB_LOG_ERROR("DBUS response error {}, {}", ec.value(),
                                      ec.message());
-                    handler(ec, satelliteInfo);
+                    handler(ec, satelliteInfo, rdeSatelliteInfo);
                     return;
                 }
 
                 // Maps a chosen alias representing a satellite BMC to a url
                 // containing the information required to create a http
                 // connection to the satellite
-                findSatelliteConfigs(objects, satelliteInfo);
+                findSatelliteConfigs(objects, satelliteInfo, rdeSatelliteInfo);
 
                 if (!satelliteInfo.empty())
                 {
@@ -847,12 +1113,18 @@ class RedfishAggregator
                         "Redfish Aggregation enabled with {} satellite BMCs",
                         std::to_string(satelliteInfo.size()));
                 }
+                else if (!rdeSatelliteInfo.empty())
+                {
+                    BMCWEB_LOG_DEBUG(
+                        "Redfish Aggregation enabled with {} RDE Device",
+                        std::to_string(rdeSatelliteInfo.size()));
+                }
                 else
                 {
                     BMCWEB_LOG_DEBUG(
                         "No satellite BMCs detected.  Redfish Aggregation not enabled");
                 }
-                handler(ec, satelliteInfo);
+                handler(ec, satelliteInfo, rdeSatelliteInfo);
             });
     }
 
diff --git a/redfish-core/lib/aggregation_service.hpp b/redfish-core/lib/aggregation_service.hpp
index be1336bf..46071487 100644
--- a/redfish-core/lib/aggregation_service.hpp
+++ b/redfish-core/lib/aggregation_service.hpp
@@ -67,7 +67,9 @@ inline void requestRoutesAggregationService(App& app)
 inline void populateAggregationSourceCollection(
     const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
     const boost::system::error_code& ec,
-    const std::unordered_map<std::string, boost::urls::url>& satelliteInfo)
+    const std::unordered_map<std::string, boost::urls::url>& satelliteInfo,
+    const std::unordered_map<std::string, RdeSatelliteConfig>& rdeSatelliteInfo
+    [[maybe_unused]])
 {
     // Something went wrong while querying dbus
     if (ec)
@@ -139,7 +141,9 @@ inline void populateAggregationSource(
     const std::string& aggregationSourceId,
     const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
     const boost::system::error_code& ec,
-    const std::unordered_map<std::string, boost::urls::url>& satelliteInfo)
+    const std::unordered_map<std::string, boost::urls::url>& satelliteInfo,
+    const std::unordered_map<std::string, RdeSatelliteConfig>& rdeSatelliteInfo
+    [[maybe_unused]])
 {
     asyncResp->res.addHeader(
         boost::beast::http::field::link,
-- 
2.43.0

